Rohan's review (#23)
cbor-wg/draft-ietf-cbor-serialization: draft-ietf-cbor-serialization.md

     date: July, 2021
     target: https://craftinginterpreters.com/optimization.html#nan-boxing

   UML:
     title: OMG Unified Modeling Language (OMG UML) Version 2.5.1
     date: December, 2017
     target: https://www.omg.org/spec/UML/2.5.1/PDF

--- abstract

This document defines two CBOR serializations: "ordinary serialization" and "deterministic serialization."

> It also introduces the term "general serialization" to name the full, variable set of serialization options defined in RFC 8949.

Author: No, as mentioned, neither the Abstract nor Section headings are allowed to contain references

General serialization consists of all of these:

* Any length CBOR argument (e.g., the integer 0 may be encoded as 0x00, 0x1800 or or 0x190000 and so on).
* Any length floating point value regardless of encoding size (e.g. 0.00 can be 0xf900, 0xfa000000000 and so on).
* Both definite or indefinite-length strings, arrays and maps are allowed.
* Big numbers can represent values that are also representable by major types 0 and 1 (e.g., 0 can be encoded as a big number, as 0xc34100).

A decoder that supports general serialization is able to decode all of these.

If a CBOR-based protocol specification does not explicitly specify serialization, general serialization is implied.
This means that a compliant decoder for such a protocol is required to accept all forms allowed by general serialization including both definite and indefinite lengths.
For example, CBOR Web Token (CWT), {{-CWT}} does not specify serialization; therefore, a full and proper CWT decoder must be able to handle variable-length CBOR arguments plus indefinite-length strings, arrays and maps.

> > Note that CBOR Object Signing and Encryption (COSE) {{-COSE}} uses effectively ordinary encoding (see {{OrdinarySerialization}}) for its internal structures, such as the `Sig_structure`.

Reviewer: Section 9 of RFC 9052 references section 4.2.1 of RFC 8949 which is about deterministic serialization. In principle COSE, absolutely requires determinism for the three structures.

Typically, there are no maps in these structures. When there are no maps ordinary == deterministic. However there can be maps in the values of parameters. For example the 'crit' header parameter is a map.

I don't think COSE should be mentioned in the context of ordinary serialization at all.

   * Big numbers described in {{Section 3.4.3 of -cbor}} MUST be accepted.
   * Leading zeros SHOULD be ignored.

>    * An empty bstr SHOULD be accepted and treated as the value zero.

Reviewer: Should be "byte string". RFC 8949 doesn't mention "bstr"

The easy implementation and broad usefulness makes ordinary serialization the best choice for most CBOR protocols.
To some degree it is a de facto standard for common CBOR protocols.

> However, it is not suitable if determinism is needed and the CBOR contains maps, because the order of items in a map is allowed to vary.

Author: I think that is just ignoring reality. The example with COSE clearly shows this. It's also the case for most of MIMI Content. Anything else designed with structs (arrays) instead of maps will also have this consideration. I don't think you can say "because the order of items in a map is allowed to vary" and not acknowledge that if there are no maps there is no such variation.

1. All of ordinary serialization defined in {{OrdinaryEncoding}} MUST be used.

1. If a map is encoded, the items in it MUST be sorted in the bytewise lexicographic order of their deterministic encodings of the map keys.
   (Note that this is the same as the sorting in {{Section 4.2.1 of -cbor}} and not the same as {{Section 3.9 of RFC7049}} / {{Section 4.2.3 of -cbor}}.

## Decoder Requirements {#DeterministicDecoding}

1. Decoders MUST meet the decoder requirements described in {{OrdinaryDecoding}}.
That is, deterministic encoding imposes no requirements over and above the requirements for decoding ordinary serialization.

> Note that a decoder MAY reject out-of-order maps as a consequence of {{SerializationChecking}}.

> 

Author: It is a Note, and it only contains a MAY. I think it is harmless and it will prevent people from saying "but the spec didn't say I was allowed to do this", which happens a LOT.

 |  | Information Model | Data Model | Serialization |
 | Abstraction Level | Top level; conceptual | Realization of information in data structures and data types | Actual bytes encoded for transmission |
 | Example | The temperature of something | A floating-point number representing the temperature | Encoded CBOR of a floating-point number |

>  | Standards | {{UML}} | CDDL | CBOR |

Author: I'm pretty sure.

- Values cannot be translated directly to JSON, which does not support NaNs of any kind.

## Clarification of RFC 8949

This is a clarifying restatement of how NaNs are treated according to {{-cbor}}.

CBOR floating point values are based on 64-bit floating point values.
NaNs represented in floating-point values of different lengths are considered equivalent by CBOR in the basic generic data model if a) their sign bits are identical, and b) their significands are identical after both significands are zero-extended on the right to 64 bits

> A NaN encoded as half-, single-, or double-precision is equivalent in CBOR whenever it satisfies the rules above.

Reviewer: The statement about equivalence in the data model is critical and can't be removed.

This section is distinct from the Core Deterministic Encoding Requirements and represents an optional alternative for handling NaNs.

## Divergence from RFC 8949 {#NaNCompatibility}

Orindary and deterministic serialization defined in this document diverge from the preferred serialization requirement in {{-cbor}} for shortest-length encoding of NaNs in ordinary and deterministic serializations.

In ordinary and deterministic serializations, non-trivial NaNs are not allowed.

> The treatment of non-trivial NaNs is the only aspect of deterministic serialization that is different from {{Section 4.2.1 of -cbor}}.

Reviewer: How about this:
The treatment of non-trivial NaNs is the sole difference between deterministic serialization and {{Section 4.2.1 of -cbor}}.

The divergence is justified by the following:

- Encoding and equivalence of non-trivial NaNs was a little unclear {{-cbor}}.
- IEEE 754 doesn't set requirements for their handling.
- Non-trivial NaNs are not well-supported across CPUs and programming environments.

> - Implementing preferred serialization for non-trivial NaNs is complex and error-prone; many CBOR implementations don't encode and/or decode non-trivial NaN, or don't encode or decode them correctly.

Reviewer: How about this:
Because preferred serialization of non-trivial NaNs is difficult and error-prone to implement, many CBOR libraries either do not support it or implement it incorrectly.

Author: I wanted to keep the part about encoding and decoding. Is this better?

```suggestion
- Because preferred serialization of non-trivial NaNs is difficult and error-prone to implement, many CBOR implementations don't encode and/or decode non-trivial NaN, or don't encode or decode them correctly.
```

While non-trivial NaNs are excluded from ordinary and deterministic serialization, they are theoretically supported by {{-cbor}}.
General serialization does support them.

> New protocol designs SHOULD avoid non-trivial NaNs.

Author: why shouldn't new protocol designs avoid non-trivial NaNs?
